### PR TITLE
Refactored of passing frame attribute/context information

### PR DESCRIPTION
**Pull request by bradfrizzell**
Previously, all of the context information used to build window.context was passed on the hash of the src url. This is being switched to the name element instead of iframes. This is done to make passing the context information down to nested iframes much more simple, specifically in relation to the AmpContext library.

### CODE REVIEW

**Comment by bradfrizzell** [comment]
apologies for the duplicate commit messages, not exactly sure what happened there. 

@lannka

File: src/3p-frame.js [hunk 1]
```diff
@@ -135,7 +142,7 @@ export function getIframe(parentWindow, parentElement, opt_type, opt_context) {
     this.readyState = 'complete';
   };
   iframe.setAttribute(
-      'data-amp-3p-sentinel', attributes._context.amp3pSentinel);
+      'data-amp-3p-sentinel', attributes._context.sentinel);
```
Comment (lannka) [inline src/3p-frame.js]: should we just generate the sentinel here, if this is the only place uses the sentinel.

File: src/3p-frame.js [hunk 2]
```diff
@@ -119,8 +121,13 @@ export function getIframe(parentWindow, parentElement, opt_type, opt_context) {
   const baseUrl = getBootstrapBaseUrl(parentWindow);
   const host = parseUrl(baseUrl).hostname;
   // Pass ad attributes to iframe via the fragment.
-  const src = baseUrl + '#' + JSON.stringify(attributes);
-  const name = host + '_' + attributes.type + '_' + count[attributes.type]++;
```
Comment (lannka) [inline src/3p-frame.js]: was the name attribute used anywhere?

File: test/functional/test-impression.js [hunk 3]
```diff
@@ -24,7 +24,7 @@ import {viewerForDoc} from '../../src/viewer';
 import {xhrFor} from '../../src/xhr';
 import * as sinon from 'sinon';
 
-describe('impression', () => {
+describe('test-impression', () => {
```
Comment (lannka) [inline test/functional/test-impression.js]: you can revert this change

File: 3p/integration.js [hunk 4]
```diff
@@ -372,6 +377,7 @@ window.draw3p = function(opt_configCallback, opt_allowed3pTypes,
     window.context.reportRenderedEntityIdentifier =
         reportRenderedEntityIdentifier;
     window.context.computeInMasterFrame = computeInMasterFrame;
+    window.context.addContextToIframe = getAddContextToIframe(iframeName);
```
Comment (lannka) [inline 3p/integration.js]: we normally use `bind` to create a function:
```
window.context.addContextToIframe = addNameToIframe.bind(null, iframeName)

function addNameToIframe(name, iframe) {
  iframe.name = name;
}
```

File: 3p/integration.js [hunk 5]
```diff
@@ -138,9 +138,14 @@ const AMP_EMBED_ALLOWED = {
...
-const data = parseFragment(location.hash);
-window.context = data._context;
-
+const iframeName = window.name;
+let data;
+try {
+  data = JSON.parse(window.name).attributes;
+  window.context = data._context;
+} catch (err) {
+  window.context = {};
```
Comment (lannka) [inline 3p/integration.js]: report dev error here

File: 3p/integration.js [hunk 6]
```diff
@@ -138,9 +138,14 @@ const AMP_EMBED_ALLOWED = {
   zergnet: true,
 };
 
-const data = parseFragment(location.hash);
-window.context = data._context;
-
+const iframeName = window.name;
+let data;
+try {
+  data = JSON.parse(window.name).attributes;
```
Comment (lannka) [inline 3p/integration.js]: iframeName

File: src/3p-frame.js [hunk 7]
```diff
@@ -119,8 +121,13 @@ export function getIframe(parentWindow, parentElement, opt_type, opt_context) {
...
+  const src = baseUrl;
```
Comment (lannka) [inline src/3p-frame.js]: this line can be removed?

File: src/3p-frame.js [hunk 8]
```diff
@@ -80,11 +80,13 @@ function getFrameAttributes(parentWindow, element, opt_type, opt_context) {
     mode: getModeObject(),
     canary: !!(parentWindow.AMP_CONFIG && parentWindow.AMP_CONFIG.canary),
     hidden: !viewer.isVisible(),
-    amp3pSentinel: generateSentinel(parentWindow),
+    sentinel: generateSentinel(parentWindow),
     initialIntersection: element.getIntersectionChangeEntry(),
     domFingerprint: domFingerprint(element),
     startTime,
   };
+  attributes.ampcontextVersion = (getMode().localDev ? 'LOCAL' :
```
Comment (lannka) [inline src/3p-frame.js]: should not be in attributes?

File: 3p/integration.js (hunk 6 above)
Comment (lannka) [inline 3p/integration.js]: probably it's a good time to revisit the data structure, it's quite confusing now.

We can change it from
```
{
  ...
  attributes: {
    _context: {}
  }
}
```

to
{
  attributes: {},
  context: {},
}

File: src/3p-frame.js (hunk 2 above)
Comment (bradfrizzell) [inline src/3p-frame.js]: we decided to make a breaking change here. 
@keithwrightbos

Comment (bradfrizzell) [inline src/3p-frame.js]: we decided to make a breaking change here @keithwrightbos

Comment (lannka) [inline src/3p-frame.js]: I was wondering if this is not used anywhere, can we just remove them?

File: test/functional/test-impression.js (hunk 3 above)
Comment (bradfrizzell) [inline test/functional/test-impression.js]: done

File: src/3p-frame.js (hunk 8 above)
Comment (lannka) [inline src/3p-frame.js]: depend on where do you need this? it's not from element attribute, and you don't want them to be shown as attribute, right?

File: 3p/integration.js (hunk 6 above)
Comment (bradfrizzell) [inline 3p/integration.js]: done

File: 3p/integration.js (hunk 4 above)
Comment (bradfrizzell) [inline 3p/integration.js]: done

File: 3p/integration.js [hunk 9]
```diff
@@ -377,7 +377,7 @@ window.draw3p = function(opt_configCallback, opt_allowed3pTypes,
     window.context.reportRenderedEntityIdentifier =
         reportRenderedEntityIdentifier;
     window.context.computeInMasterFrame = computeInMasterFrame;
-    window.context.addContextToIframe = getAddContextToIframe(iframeName);
+    window.context.addContextToIframe = addNameToIframe.bind(null, iframeName);
```
Comment (jridgewell) [inline 3p/integration.js]: This is bound to the `iframe` param now, not the name.

Comment (bradfrizzell) [inline 3p/integration.js]: fixed

File: 3p/integration.js [hunk 10]
```diff
@@ -485,6 +493,18 @@ function reportRenderedEntityIdentifier(entityId) {
 }
 
 /**
+ *  Adds the serialized ad attributes to an iframe's name attribute.
+ *  @param {Iframe} iframe
```
Comment (jridgewell) [inline 3p/integration.js]: Need to swap.

File: src/3p-frame.js [hunk 11]
```diff
@@ -80,11 +80,13 @@ function getFrameAttributes(parentWindow, element, opt_type, opt_context) {
     mode: getModeObject(),
     canary: !!(parentWindow.AMP_CONFIG && parentWindow.AMP_CONFIG.canary),
     hidden: !viewer.isVisible(),
-    amp3pSentinel: generateSentinel(parentWindow),
+    sentinel: generateSentinel(parentWindow),
```
Comment (jridgewell) [inline src/3p-frame.js]: Is this a breaking change?

File: 3p/integration.js [hunk 12]
```diff
@@ -138,9 +138,16 @@ const AMP_EMBED_ALLOWED = {
...
 };
 
-const data = parseFragment(location.hash);
-window.context = data._context;
-
+const iframeName = window.name;
+let data;
+try {
+  data = JSON.parse(iframeName).attributes;
+  window.context = data._context;
```
Comment (jridgewell) [inline 3p/integration.js]: This is now at `data.attributes.context`, per https://github.com/ampproject/amphtml/pull/6498/files#diff-34e84f9cf5e8134d86bf1fc893ab293aR128.

File: src/3p-frame.js [hunk 13]
```diff
@@ -119,12 +121,16 @@ export function getIframe(parentWindow, parentElement, opt_type, opt_context) {
...
+  const name = JSON.stringify({
+    host,
+    type: attributes.type,
+    count: count[attributes.type],
+    attributes,
+  });
 
-  iframe.src = src;
+  iframe.src = baseUrl;
   iframe.name = name;
```
Comment (jridgewell) [inline src/3p-frame.js]: Need to `decodeURIComponent`?

File: src/3p-frame.js (hunk 2 above)
Comment (bradfrizzell) [inline src/3p-frame.js]: I believe it is used, at least there are tests that check for it

File: src/3p-frame.js [hunk 14]
```diff
@@ -119,12 +121,16 @@ export function getIframe(parentWindow, parentElement, opt_type, opt_context) {
   const baseUrl = getBootstrapBaseUrl(parentWindow);
   const host = parseUrl(baseUrl).hostname;
   // Pass ad attributes to iframe via the fragment.
-  const src = baseUrl + '#' + JSON.stringify(attributes);
-  const name = host + '_' + attributes.type + '_' + count[attributes.type]++;
+  const name = JSON.stringify({
+    host,
+    type: attributes.type,
+    count: count[attributes.type],
```
Comment (jridgewell) [inline src/3p-frame.js]: Need to increment this value (this was a hard fought bug).

File: 3p/integration.js (hunk 12 above)
Comment (jridgewell) [inline 3p/integration.js]: Ohhh! I missed the `.attributes` on the line above. Ignore this comment.

File: 3p/integration.js (hunk 6 above)
Comment (lannka) [inline 3p/integration.js]: in case this comment was buried.

File: src/3p-frame.js (hunk 14 above)
Comment (bradfrizzell) [inline src/3p-frame.js]: done

File: src/3p-frame.js (hunk 13 above)
Comment (bradfrizzell) [inline src/3p-frame.js]: no

**Comment by bradfrizzell** [comment]
ptal

**Comment by lannka** [comment]
Some of my comments was buried. For example: https://github.com/ampproject/amphtml/pull/6498#discussion_r91761447
https://github.com/ampproject/amphtml/pull/6498#discussion_r91763419

File: 3p/integration.js (hunk 6 above)
Comment (bradfrizzell) [inline 3p/integration.js]: I can fix this, but it would be a lot easier for me to do it in #6506 if that's alright with you. I need to make a helper function that constructs the whole name object anyways, and that's the PR where I'm going to put it

File: src/3p-frame.js (hunk 8 above)
Comment (bradfrizzell) [inline src/3p-frame.js]: I'll move this as part of the #6506 as I am already making the name building helper function there

File: 3p/integration.js (hunk 5 above)
Comment (bradfrizzell) [inline 3p/integration.js]: done

**Comment by bradfrizzell** [comment]
Sorry about that, I addressed them. Changing the whole data structure of the attributes would be easier for me to fix in #6506 if that's alright. If you think it must be done as part of this PR I can do that

File: src/3p-frame.js (hunk 14 above)
Comment (lannka) [inline src/3p-frame.js]: @jridgewell do you know what is the counter used for? I don't see it documented anywhere.

Comment (jridgewell) [inline src/3p-frame.js]: The counter is to ensure the iframe's window is not cached. https://github.com/ampproject/amphtml/pull/2955

Comment (lannka) [inline src/3p-frame.js]: Ah, good to know. @bradfrizzell could you add a link in the code?

**Comment by lannka** [comment]
Ah, there's another previous comment not addressed yet (about the master selection code): 
https://github.com/ampproject/amphtml/pull/6498#pullrequestreview-12270017

**Comment by bradfrizzell** [comment]
ptal @lannka

File: 3p/integration.js [hunk 15]
```diff
@@ -519,6 +529,16 @@ function reportRenderedEntityIdentifier(entityId) {
 }
 
 /**
+ *  Adds the serialized ad attributes to an iframe's name attribute.
+ *  @param {string} name A URIencoded JSON object of context
+ *      that needs to be attached to an iframe.
+ *  @param {Element} iframe
+ */
+function addNameToIframe(name, iframe) {
```
Comment (lannka) [inline 3p/integration.js]: hmm, seems now it's only used in one place, shall we just inline the function?

File: 3p/integration.js [hunk 16]
```diff
@@ -152,9 +152,18 @@ const AMP_EMBED_ALLOWED = {
   zergnet: true,
 };
 
-const data = parseFragment(location.hash);
-window.context = data._context;
-
+const iframeName = window.name;
```
Comment (lannka) [inline 3p/integration.js]: please explain why we cache this value.

File: src/3p-frame.js [hunk 17]
```diff
@@ -121,6 +121,11 @@ export function getIframe(parentWindow, parentElement, opt_type, opt_context) {
   const baseUrl = getBootstrapBaseUrl(parentWindow);
   const host = parseUrl(baseUrl).hostname;
   // Pass ad attributes to iframe via the fragment.
+
+  // This name attribute may be overwritten if this frame is chosen to
+  // be the master frame. That is ok, as we will read the name off
+  // for our uses before that would occur.
+  // @see https://github.com/ampproject/amphtml/blob/master/3p/integration.js#L315
```
Comment (lannka) [inline src/3p-frame.js]: nit: remove the line number as it might soon be irrelevant

File: 3p/integration.js (hunk 16 above)
Comment (bradfrizzell) [inline 3p/integration.js]: don't need to, will fix

Comment (bradfrizzell) [inline 3p/integration.js]: I suppose we don't need to. fixed

File: src/3p-frame.js (hunk 17 above)
Comment (bradfrizzell) [inline src/3p-frame.js]: done

**Comment by bradfrizzell** [comment]
ptal @lannka

File: 3p/integration.js [hunk 18]
```diff
@@ -519,6 +529,14 @@ function reportRenderedEntityIdentifier(entityId) {
 }
 
 /**
+ *  Adds the serialized ad attributes to an iframe's name attribute.
+ *  @param {string} name A URIencoded JSON object of context
+ *      that needs to be attached to an iframe.
+ *  @param {Element} iframe
+ */
+function addNameToIframe(name, iframe) {iframe.name = name;}
```
Comment (lannka) [inline 3p/integration.js]: sorry, by inline i meant to have
```
window.context.addContextToIframe = (iframe, name) => { 
  iframe.name = name; 
}
```

File: 3p/integration.js (hunk 16 above)
Comment (lannka) [inline 3p/integration.js]: i actually meant to add a comment why we cache this value.
Otherwise, you might want to move master selection code below all the code that read window.name.

Comment (bradfrizzell) [inline 3p/integration.js]: d'oh. undid and added comment

File: 3p/integration.js (hunk 18 above)
Comment (lannka) [inline 3p/integration.js]: oh sorry,  i meant
```
window.context.addContextToIframe = iframe => { 
  iframe.name = iframeName; 
}
```

**Comment by bradfrizzell** [comment]
ptal

**Comment by bradfrizzell** [comment]
@lannka I added an experiment flag. Please take a look, let me know if I did this correctly.

File: 3p/messaging.js [hunk 19]
```diff
@@ -25,7 +25,7 @@ export function nonSensitiveDataPostMessage(type, opt_object) {
   }
   const object = opt_object || {};
   object.type = type;
-  object.sentinel = window.context.amp3pSentinel;
+  object.sentinel = window.context.sentinel;
```
Comment (lannka) [inline 3p/messaging.js]: could you make this naming change in a separate PR? In case this renaming break anything, we can easily revert a smaller PR.

File: 3p/integration.js [hunk 20]
```diff
@@ -131,6 +132,21 @@ import {zergnet} from '../ads/zergnet';
 import {zucks} from '../ads/zucks';
 
 /**
+ * If true, then in experiment where the passing of context metadata
+ * has been moved from the iframe src hash to the iframe name attribute.
+ */
+const nameExpOn = isExperimentOn(window, 'move-context-to-name');
```
Comment (lannka) [inline 3p/integration.js]: the flag value is not available inside iframe. (I'm working on it right now #6684 ). To not be blocked by me, I suggest you read from fragment first, and then from name.

File: src/3p-frame.js [hunk 21]
```diff
@@ -27,6 +28,11 @@ import {urls} from './config';
 import {setStyle} from './style';
 import {domFingerprint} from './utils/dom-fingerprint';
 
+/**
+ * If true, then in experiment where the passing of context metadata
+ * has been moved from the iframe src hash to the iframe name attribute.
+ */
+const nameExpOn = isExperimentOn(self, 'move-context-to-name');
```
Comment (lannka) [inline src/3p-frame.js]: suggested naming: '3p-iframe-context-in-name'

File: src/3p-frame.js (hunk 8 above)
Comment (lannka) [inline src/3p-frame.js]: seems this line can be added in your next PR, if you plan to move it eventually

File: 3p/messaging.js (hunk 19 above)
Comment (bradfrizzell) [inline 3p/messaging.js]: ok

File: 3p/integration.js (hunk 20 above)
Comment (bradfrizzell) [inline 3p/integration.js]: done

File: src/3p-frame.js (hunk 21 above)
Comment (bradfrizzell) [inline src/3p-frame.js]: ended up using 'iframeContextInName'

**Comment by bradfrizzell** [comment]
addressed your comments @lannka

File: tools/experiments/experiments.js [hunk 22]
```diff
@@ -231,6 +231,11 @@ const EXPERIMENTS = [
     name: 'AMP Accordion attribute to opt out of preserved state.',
     Spec: 'https://github.com/ampproject/amphtml/issues/3813',
   },
+  {
+    id: 'move-context-to-name',
```
Comment (lannka) [inline tools/experiments/experiments.js]: Change this `move-context-to-name` to `3p-frame-context-in-name`?

File: tools/experiments/experiments.js [hunk 23]
```diff
@@ -231,6 +231,11 @@ const EXPERIMENTS = [
...
+    name: 'Move passing context metadata from url hash to name attribute',
+    Spec: '',
+  },
```
Comment (lannka) [inline tools/experiments/experiments.js]: please create a `cleanupIssue` as well

File: 3p/integration.js [hunk 24]
```diff
@@ -143,8 +143,25 @@ const AMP_EMBED_ALLOWED = {
   zergnet: true,
 };
 
-const data = parseFragment(location.hash);
-window.context = data._context;
+let data;
+// Need to cache iframeName as it will be potentially overwritten by
+// masterSelection, as per below.
+const iframeName = window.name;
+try {
```
Comment (lannka) [inline 3p/integration.js]: can you move the try-catch into `parseFragment`? And make it return `null` for exception.

File: tools/experiments/experiments.js (hunk 23 above)
Comment (bradfrizzell) [inline tools/experiments/experiments.js]: done

File: 3p/integration.js (hunk 24 above)
Comment (bradfrizzell) [inline 3p/integration.js]: done

**Comment by bradfrizzell** [comment]
Checked those pages. they all looked good. I'm going to make sure the few console messages are all the same on the master branch

**Comment by bradfrizzell** [comment]
gah sorry about the duplicated commits. I just stupidly rebased without thinking

**Comment by bradfrizzell** [comment]
ptal / merge if ready

File: src/3p-frame.js [hunk 25]
```diff
@@ -27,6 +28,11 @@ import {urls} from './config';
 import {setStyle} from './style';
 import {domFingerprint} from './utils/dom-fingerprint';
 
+/**
+ * If true, then in experiment where the passing of context metadata
+ * has been moved from the iframe src hash to the iframe name attribute.
+ */
+const iframeContextInName = isExperimentOn(self, 'move-context-to-name');
```
Comment (lannka) [inline src/3p-frame.js]: the string literal needs to be identical to experiment ID you defined in `experiments.js`.
So, `'move-context-to-name'` => `'3p-frame-context-in-name'`

File: tools/experiments/experiments.js [hunk 26]
```diff
@@ -231,6 +231,12 @@ const EXPERIMENTS = [
     name: 'AMP Accordion attribute to opt out of preserved state.',
     Spec: 'https://github.com/ampproject/amphtml/issues/3813',
   },
+  {
+    id: 'iframeContextInName',
```
Comment (lannka) [inline tools/experiments/experiments.js]: to follow the convention: '3p-frame-context-in-name'

File: src/3p-frame.js (hunk 25 above)
Comment (bradfrizzell) [inline src/3p-frame.js]: ah whoops. fixed.

File: tools/experiments/experiments.js (hunk 26 above)
Comment (bradfrizzell) [inline tools/experiments/experiments.js]: done

**Comment by bradfrizzell** [comment]
Ran the manual tests both ways. Everything looked fine, no new error messages/differences between experiment on/off.

**Comment by bradfrizzell** [comment]
PTAL

File: test/functional/test-3p-frame.js [hunk 27]
```diff
@@ -155,11 +163,17 @@ describe('3p-frame', () => {
     expect(locationHref).to.not.be.empty;
     const docInfo = documentInfoForDoc(window.document);
     expect(docInfo.pageViewId).to.not.be.empty;
-    const amp3pSentinel = iframe.getAttribute('data-amp-3p-sentinel');
+    let amp3pSentinel;
+    if (iframeContextInName) {
```
Comment (jridgewell) [inline test/functional/test-3p-frame.js]: These tests need to be done with both `iframeContextInName` set to false and set to true. As it is, we'll never test the `true` case.
